From 14fedcf2e69fd11ba682cc8721d537f0278b8b1c Mon Sep 17 00:00:00 2001
From: cheney <11722200@qq.com>
Date: Wed, 6 Aug 2025 09:53:39 +0800
Subject: [PATCH 432/540] Changelogs: 1. cpufreq control delete boost_1_6ghz
 flag; 2. CLK_FREQ_1600M/CLK_FREQ_1500M voltage change to 0.9v default; 3.
 delete 1.2G and 1.3G frequency because impacting on the stablility of d2d

---
 .../boot/dts/eswin/eic770x-opptable.dtsi      | 31 +++----
 drivers/clk/eswin/clk.c                       | 82 +++++++++----------
 drivers/clk/eswin/clk.h                       |  1 +
 3 files changed, 51 insertions(+), 63 deletions(-)

diff --git a/arch/riscv/boot/dts/eswin/eic770x-opptable.dtsi b/arch/riscv/boot/dts/eswin/eic770x-opptable.dtsi
index 6198424e38d5e..5dd847e8bf93a 100644
--- a/arch/riscv/boot/dts/eswin/eic770x-opptable.dtsi
+++ b/arch/riscv/boot/dts/eswin/eic770x-opptable.dtsi
@@ -27,6 +27,7 @@
 #define CLK_FREQ_1400M					1400000000
 #define CLK_FREQ_1300M					1300000000
 #define CLK_FREQ_1200M					1200000000
+#define CLK_FREQ_1100M					1100000000
 #define CLK_FREQ_1000M					1000000000
 #define CLK_FREQ_900M					900000000
 #define CLK_FREQ_800M					800000000
@@ -93,16 +94,11 @@ opp-1000000000 {
 			opp-microvolt = <800000>;
 			clock-latency-ns = <70000>;
 		};
-		opp-1200000000 {
-			opp-hz = /bits/ 64 <CLK_FREQ_1200M>;
-			opp-microvolt = <800000>;
-			clock-latency-ns = <70000>;
-		};
-		opp-1300000000 {
-			opp-hz = /bits/ 64 <CLK_FREQ_1300M>;
-			opp-microvolt = <800000>;
-			clock-latency-ns = <70000>;
-		};
+		opp-1100000000 {
+			opp-hz = /bits/ 64 <CLK_FREQ_1100M>;
+ 			opp-microvolt = <800000>;
+ 			clock-latency-ns = <70000>;
+ 		};
 		opp-1400000000 {
 			opp-hz = /bits/ 64 <CLK_FREQ_1400M>;
 			opp-microvolt = <800000>;
@@ -164,16 +160,11 @@ opp-1000000000 {
 			opp-microvolt = <800000>;
 			clock-latency-ns = <70000>;
 		};
-		opp-1200000000 {
-			opp-hz = /bits/ 64 <CLK_FREQ_1200M>;
-			opp-microvolt = <800000>;
-			clock-latency-ns = <70000>;
-		};
-		opp-1300000000 {
-			opp-hz = /bits/ 64 <CLK_FREQ_1300M>;
-			opp-microvolt = <800000>;
-			clock-latency-ns = <70000>;
-		};
+		opp-1100000000 {
+			opp-hz = /bits/ 64 <CLK_FREQ_1100M>;
+ 			opp-microvolt = <800000>;
+ 			clock-latency-ns = <70000>;
+ 		};
 		opp-1400000000 {
 			opp-hz = /bits/ 64 <CLK_FREQ_1400M>;
 			opp-microvolt = <800000>;
diff --git a/drivers/clk/eswin/clk.c b/drivers/clk/eswin/clk.c
index 8e1fc15c44346..9355f48adafbc 100644
--- a/drivers/clk/eswin/clk.c
+++ b/drivers/clk/eswin/clk.c
@@ -35,11 +35,9 @@
 #define NUMA_NODE_NUMS 2
 
 struct cpu_info {
-	bool cpu_no_boost_1_6ghz;
 	uint64_t cpu_freqhz[NUMA_NODE_NUMS];
 };
-struct cpu_info g_cpu_info = {.cpu_no_boost_1_6ghz = true,
-								.cpu_freqhz = {0}};
+struct cpu_info g_cpu_info = {.cpu_freqhz = {0}};
 static DEFINE_MUTEX(lock);
 
 struct clk_hw *eswin_clk_find_parent(struct eswin_clock_data *data,
@@ -223,6 +221,12 @@ static int eswin_calc_pll(u32 *frac_val, u32 *postdiv1_val, u32 *fbdiv_val,
 			*fbdiv_val = 200;
 			*refdiv_val = 1;
 			break;
+		case CLK_FREQ_1100M:
+			*frac_val = 5592405;
+			*postdiv1_val = 0;
+			*fbdiv_val = 183;
+			*refdiv_val = 1;
+			break;
 		case CLK_FREQ_1000M:
 			*frac_val = 11184810;
 			*postdiv1_val = 0;
@@ -295,14 +299,6 @@ static int eswin_calc_pll(u32 *frac_val, u32 *postdiv1_val, u32 *fbdiv_val,
 	return ret;
 }
 
-static bool cpu_no_boost_1_6ghz;
-static int __init cpu_no_boost_1_6ghz_setup(char *__unused)
-{
-	cpu_no_boost_1_6ghz = true;
-	return 1;
-}
-__setup("cpu_no_boost_1_6ghz", cpu_no_boost_1_6ghz_setup);
-
 #define to_pll_clk(_hw) container_of(_hw, struct eswin_clk_pll, hw)
 static int clk_pll_set_rate(struct clk_hw *hw, unsigned long rate,
 			    unsigned long parent_rate)
@@ -320,6 +316,7 @@ static int clk_pll_set_rate(struct clk_hw *hw, unsigned long rate,
 	char clk_cpu_lp_pll_name[50] = { 0 };
 	char clk_cpu_pll_name[50] = { 0 };
 	enum voltage_level cpu_target_volatge = VOLTAGE_0_8V;
+	unsigned long max_rate;
 
 	ret = eswin_calc_pll(&frac_val, &postdiv1_val, &fbdiv_val, &refdiv_val,
 			     (u64)rate, clk);
@@ -381,52 +378,49 @@ static int clk_pll_set_rate(struct clk_hw *hw, unsigned long rate,
 		}
 		mutex_lock(&lock);
 		g_cpu_info.cpu_freqhz[clk->numa_id] = rate;
-		rate = g_cpu_info.cpu_freqhz[0] > g_cpu_info.cpu_freqhz[1] ? 
+		max_rate = g_cpu_info.cpu_freqhz[0] > g_cpu_info.cpu_freqhz[1] ? 
 				g_cpu_info.cpu_freqhz[0] : g_cpu_info.cpu_freqhz[1];
 		/*
 		 * The CPU clock has now switched to the LP_PLL,
-		 * so we can adjust the CPU's supply voltage
-		 * If the board cpu voltage does not support boosting to 0.9V,
+		 * so we can adjust the CPU's supply volatge
+		 * If the board cpu volatge does not support boosting to 0.9V,
 		 * then the frequency cannot exceed 1.6GHz.
 		 */
-		switch (rate) {
+		switch (max_rate) {
 		case CLK_FREQ_1800M:
 		case CLK_FREQ_1700M:
-			cpu_target_volatge = VOLTAGE_0_9V;
-			break;
 		case CLK_FREQ_1600M:
 		case CLK_FREQ_1500M:
-			cpu_target_volatge = true == g_cpu_info.cpu_no_boost_1_6ghz ?
-						     VOLTAGE_0_8V :
-						     VOLTAGE_0_9V;
+			cpu_target_volatge = VOLTAGE_0_9V;
 			break;
 		default:
 			cpu_target_volatge = VOLTAGE_0_8V;
 			/*
-			 * For boards that do not support voltage switching,
-			 * the voltage is maintained at 0.8V.
+			 * For boards that do not support volatge switching,
+			 * the volatge is maintained at 0.8V.
 			 * Therefore, this is also considered successful.
 			 */
 			break;
 		}
+		mutex_unlock(&lock);
 	}
-	mutex_unlock(&lock);
 	
-	ret = eswin_clk_set_cpu_volatge(clk->cpu_voltage_gpio,
-					cpu_target_volatge);
-	if (ret) {
-		pr_warn("failed to change cpu volatge to %d mV, not support rate %ld\n",
-			cpu_target_volatge, rate);
-		goto switch_back;
-	} else {
-		if (clk->cpu_current_volatge !=
-			cpu_target_volatge) {
-			pr_info("cpu volatge change to %d mV, target rate %ld\n",
-				cpu_target_volatge, rate);
-			clk->cpu_current_volatge =
-				cpu_target_volatge;
+	if (clk->cpu_current_volatge !=	cpu_target_volatge) {
+		ret = eswin_clk_set_cpu_volatge(clk->cpu_voltage_gpio,
+						cpu_target_volatge);
+		if (ret) {
+			pr_warn("cluster%d failed to change cpu to %s volatge, not support rate %ld\n",
+						clk->numa_id, cpu_target_volatge == VOLTAGE_0_9V?"high":"low", rate);
+			goto switch_back;
+		} else {
+			pr_info("cluster%d change to %s volatge, target rate %ld\n",
+					clk->numa_id, cpu_target_volatge == VOLTAGE_0_9V?"high":"low", rate);
+			clk->cpu_current_volatge = cpu_target_volatge;
 		}
-	}
+	} else {
+		pr_info("cluster%d keep %s volatge, target rate %ld\n",
+				clk->numa_id, cpu_target_volatge == VOLTAGE_0_9V?"high":"low", rate);
+ 	}
 
 	/*first disable pll */
 	val = readl_relaxed(clk->ctrl_reg0);
@@ -554,6 +548,9 @@ static unsigned long clk_pll_recalc_rate(struct clk_hw *hw,
 		case 200:
 			rate = CLK_FREQ_1200M;
 			break;
+		case 183:
+			rate = CLK_FREQ_1100M;
+			break;
 		case 166:
 			rate = CLK_FREQ_1000M;
 			break;
@@ -612,9 +609,9 @@ static long clk_pll_round_rate(struct clk_hw *hw, unsigned long rate,
 	u64 cpu_pll_clk[] = { CLK_FREQ_100M,  CLK_FREQ_200M,  CLK_FREQ_400M,
 			      CLK_FREQ_500M,  CLK_FREQ_600M,  CLK_FREQ_700M,
 			      CLK_FREQ_800M,  CLK_FREQ_900M,  CLK_FREQ_1000M,
-			      CLK_FREQ_1200M, CLK_FREQ_1300M, CLK_FREQ_1400M,
-			      CLK_FREQ_1500M, CLK_FREQ_1600M, CLK_FREQ_1700M,
-			      CLK_FREQ_1800M };
+			      CLK_FREQ_1100M, CLK_FREQ_1200M, CLK_FREQ_1300M,
+				  CLK_FREQ_1400M, CLK_FREQ_1500M, CLK_FREQ_1600M,
+				  CLK_FREQ_1700M, CLK_FREQ_1800M };
 
 	switch (clk->id) {
 	case EIC7700_APLL_FOUT1:
@@ -673,9 +670,8 @@ void eswin_clk_register_pll(struct eswin_pll_clock *clks, int nums,
 	}
 
 	mutex_lock(&lock);
-	if (g_cpu_info.cpu_no_boost_1_6ghz)
-		g_cpu_info.cpu_no_boost_1_6ghz = true == cpu_no_boost_1_6ghz;
-	g_cpu_info.cpu_freqhz[data->numa_id] = CLK_FREQ_1400M;
+	if(data->numa_id >= 0)
+		g_cpu_info.cpu_freqhz[data->numa_id] = CLK_FREQ_1400M;
 	mutex_unlock(&lock);
 
 	for (i = 0; i < nums; i++) {
diff --git a/drivers/clk/eswin/clk.h b/drivers/clk/eswin/clk.h
index d80b13427f7e6..2341d4013ab68 100644
--- a/drivers/clk/eswin/clk.h
+++ b/drivers/clk/eswin/clk.h
@@ -19,6 +19,7 @@
 #define CLK_FREQ_1400M 1400000000
 #define CLK_FREQ_1300M 1300000000
 #define CLK_FREQ_1200M 1200000000
+#define CLK_FREQ_1100M 1100000000
 #define CLK_FREQ_1000M 1000000000
 #define CLK_FREQ_900M 900000000
 #define CLK_FREQ_800M 800000000
-- 
2.53.0

